### DISCUSSION TITLE
MAF based flow bench fluctuating flow value

### DISCUSSION THREAD

**Discussion by black-top** [DIY-Flow-Bench/DIY-Flow-Bench#175] [Support]
### Version Info

V.2.0-RC.6 Build: 23041301

### Steps to replicate issue

```bash
Any time.
```

### Describe your issue

A lot of work was done since last time I wrote here. So I was fighting the same issue since the beginning.
 The problem is fluctuating flow value, and the fluctuations are more than 100cfm at stable 28"h20 pressure. 
Tested a lot of things but get the same results.

So first of all I was thinking that maybe there is leakage in the system, so I sealed everything as mush as I can
, still get the same results, then I was thinking maybe MAF gets turbulent flow so I made modifications to flow bench with 1.5M long pipes before and after flow meter,  Still same problem,

I made honeycomb structure from straws inside the pipe before maf to straiten the flow, same results, i thought may be the maf is defective, because it is original but used maf, so I bought a new aftermarket maf and got the same results,

 So I thought maybe it was EMF in the wires that caused that problem, so I used shielded wire for maf 
and separated the esp board from the vfd as far as possible, but nothing helped. 
Last thing, I made a new flow bench minimalistic setup to make sure it is leak free but nothing changed, same situation.

Don't have more ideas what can be wrong. I made a few short videos.  The plate at that test is a 22.3mm small metal disk hole, the disc is glued to the adapter plate with silicone and additionally sealed with tape to guarantee leak-free operation.
 Any ideas and suggestions? 

Videos: 
https://we.tl/t-M7qBPV7GE4

MAF Data File: BOSCH_0280218067
Ref Press Sensor: MPXV7025

### Posting Guidelines

- [x] I have read the [Posting Guidelines](https://github.com/DeeEmm/DIY-Flow-Bench/discussions/133)

**Comment by stefan63** [comment]
Measure the 5V reverence voltage as close to the MAF sensor as possible and se if it has a stable 5V feed.

**Reply by black-top** [reply]
Yes, the voltage is stable, I even tried a separate power supply with l7805 for maf with common ground, same issue.

**Comment by black-top** [comment]
I'll check tomorrow the maf signal with a scope directly on esp board, maybe some capacitance could help on adc input. But maybe there might be a problem with the program, might be somewhere something is wrong with floating point math or something else. Does anyone already use a MAF based system?

**Comment by DeeEmm** [comment]
Sorry that you are experiencing an issue, and of course a big thanks for your investigative work. This is the kind beta testing that we need on the project so that we can make it stable and universal.

If you can carry on with the hardware testing to see if you can isolate where the swing is coming from. I will try to think of some tests that we can carry out to trace the signal process through the code.

Things to look for...What frequency are the swings? Is it a stable slow swing, fast, completely random, does the value 'spike' ?

Unfortunately I could not view the vide you linked to.

I would start off by looking at each sensor input to see if there is variation on the input (in software). This may require some custom code to print the values continuously to the serial monitor. I will take a look at this.

There is already some signal conditioning on the input pin in the form of a 1uf capacitor to ground. I think this is C7 on the PCB that you have (V2.0). You can try increasing the value of this to see if it makes a difference.

At present there is no software filtering, so there is also the potential to add this, however if there are swings of 100cfm, then I suspect that there is definitely an issue to address somewhere.

You can also experiment with the Cyclical Average Buffer setting under the configuration tab. This is a simple averaging function that displays the average value over the specified number of cycles. It is the only filtering applied to the output. If the value is spiking you may be able to filter it out by increasing this value. Note that this buffer is in terms of processor cycles, so with an ESP32 processor which cycles at 80kHz, you can increase this substantially, however you need to consider that increasing this too much may overflow the buffer.

**Reply by DeeEmm** [reply]
You can also try disabling the BME and using fixed values to see if the swing is coming from the BME.

**Reply by black-top** [reply]
At the main flow bench screen, all values are stable, except flow. The depression I set with vfd, and it is moving a little bit, but I say it is stable too. Both the corrected and uncorrected flow values are jumping. Value looks fairly random.

**Reply by DeeEmm** [reply]
A random value may mean interference.

I would not expect the VFD to have that kind of impact. If we look at Poiseuille’s Law and ignore air density and pipe length the change in flow is approximately equal to the pressure differential, so you would need to be seeing the same kind of variation in the reference pressure, which I would expect would be fairly obvious if it were the cause.

I suspect that it might be a transient value and that this is getting amplified by the code. If we can identify the source, then we can add some code to filter out the spurious values and stabilise the reading.

**Reply by stefan63** [reply]
> You can also try disabling the BME and using fixed values to see if the swing is coming from the BME.

That suggestion make me think load agan ;)
Maybe there should be a "start messuring" button in the UI that will lock in the eviromental sensor value until messuring session has been finished. This will take away some values that might have effect on readout flucturations ......... iI think....

**Reply by DeeEmm** [reply]
Yes, but if the values are locked, and then the temp / baro / humidity changes in between measurements, the results will be inaccurate.

I guess we could argue 'how much will this inaccuracy be'? But then we are comparing an inaccuracy against the fluctuations, when it is possible (even likely) that the fluctuations are actually reporting accurate data. I think out of the two 'issues' that it is better to allow the fluctuations. 

The thing that we need to determine, is whether the fluctuations are expected / normal. For example as a result of bad sensor range choice.

**Comment by black-top** [comment]
I sealed flow bench set vfd about 35"h2o depression and tested for leaks, I get 0 flow value. Did the same and pressed calibrate leak test and got these values, is this correct? Leak test value -1000 does not change.  After that, I performed flow offset calibration with CD disc, I set depression to 10"h2o, and got a jumping value between 70 to 125CFM .  After calibration, I got same from 70 to 125CFM.  
![calib1](https://github.com/user-attachments/assets/24196dbf-29e8-4cbf-a566-65ae3fe1d5a0)

**Reply by DeeEmm** [reply]
hmmnnn something is not right there. The calibration offset should be the difference between the stated flow rate of the test orifice and the actual measured flow rate of the test orifice. It appears to be saving the orifice flow rate as the offset. This is definitely not correct and could be a contributing factor. 

The vac leak test looks like it is going out of range. 

The basic premise of the leak test, is that we take a measurement with a closed bench. We can then compare against this measurement by closing off the inlet of whatever we are measuring and noting the difference.

The out of range may simply be the result of the vac source exceeding the range of the pDiff sensor. 

I will take a look at that and maybe rethink how we measure the leak test.

/DM

**Reply by black-top** [reply]
Leak test, I think, is ok, because I set 35"h2o pressure with sealed system and I don't feel any air flow out of the pump exhaust, so it's probably correct that 0flow, if I increase pressure a lot, I start to see about 1-2CFM value. My vacuum sensor is MPXV7025 so it is a wider range sensor. 

For the flow fluctuations what is interesting that minimum and maximum values looks like doesn't change, it fluctuates between those values. So if it will be interference I think it will be random too. 

The 5V power supply is left from the last tests I did. It is l7805 few centimeters away from the maf connector, so the power supply is as stable as possible.

**Comment by black-top** [comment]
So I did a simple test:
CD disc sealed to the plate, temperature 16.98 °C, Barometric 100.89,  humidity 56.06%
Depression set to 9.95–10.08 'h2O   
Flow uncorrected : 80-104CFM
MAF voltage measured on the pcb : 2.96V-3.10V

Second test, everything is the same except depression, now ~28"h2o.
Depression set to 28.0-28.15 'h20  
Flow uncorrected: 120-228CFM
MAF voltage measured on PCB : 3.16V-3.55V

**Reply by black-top** [reply]
I did a lot of work today. First, I made new shielded wiring for maf with better 5V supply for maf reference voltage and proper grounding and shielding connections. I did some signal probing with an oscilloscope at different places, and I'm pretty sure that is not emf interference.  I even took it out flow element and put it directly on vfd, motor and power cables and did not see any interference. 

Next, I tested different flow straighteners, from simple mesh to honeycomb flow straighteners made from straws, and the results showed that it makes almost no difference. After that, I was thinking of putting another similar MAF  to compare readings, and while I was thinking how to do it, I was putting  some additional pipes and found a pretty interesting thing.  I tested maf sensor with 1.5m long pipe to the "measuring end"  and with 30cm pipe and got similar results, but when I puted 90deg pipe elbow after the "measuring end" I got much more stable readings even with short pipe to MAF. 
 
 I guess even with long pipes and flow straighteners, if MAF is in direct line to the measurement end, it somehow gets turbulent air, and it affects maf readings a lot. Tomorrow I will do more testing with pipes and maf position.

Probably the only problem with the code for now is that flow calibration does not work and leak calibration does not work. Also, I found that I can't set the MAF minimum voltage, after saving, it always returns to the 0.00 value.

**Reply by DeeEmm** [reply]
The distance that the flow takes to settle from turbulent to laminar (fully developed) can be calculated but it is generally too excessive to be implemented in a regular flow bench. IIRC it will be somewhere in the region of 4 metre for a 100mm diameter pipe (40D IIRC). On lab based MAF measurement and calibration systems you will often see the MAF mounted in the middle of a very long straight manifold.

For this reason, on shorter systems flow straighteners are employed. Generally the straighteners built into the MAF sensors work relatively well. Or at least well enough. IIRC another member tried various experiments with flow straighteners using straws etc but the results were not as expected

The results using a 90 degree bend do not surprise me. Using a 90 degree bend is a common technique used in turbo systems to translate the cyclone style circular flow into linear flow. If this gives the most stable readings and can be integrated into your bench design, then it might be an appropriate solution.

**Reply by black-top** [reply]
So I did some work today. Connected pipe with a few 90-degree bends, and it gave a huge difference in CFM fluctuations, now it fluctuates only 2-5CFM.
The CD flows  16-18CFM  at 10"h2o, so not that far off. 
Today I noticed that the calibration offset value after the esp board is restarted goes back to 1000 each time. Is it suppose to do that? It doesn't save offset value to memory?  Leak calibration value is always -1000 and MAF Min volts:  always stays 0.00, can't be changed.
Another small issue I noticed today is that if the vacuum or flow value changes rapidly, for example, after disconnecting the vacuum hose from the sensor, the user interface screen freezes, and it takes a few minutes to recover, or sometimes it recovers only after restarting the esp32 board. I have tested different refresh rates and buffer values, but it looks like the problem is 
somewhere else.

**Reply by DeeEmm** [reply]
Do you know if the serial monitor is still working at this point? If there was a panic the resultant error message would be displayed on the serial monitor. You can also see if it the serial monitor is still working by issuing a command to it. 

If the issue is WiFi related you can try to recover the WiFi connection by issuing the '$' command to the serial monitor. This will try to re-establish the WiFi connection

If it is going into panic I will need to take a look at that. Perhaps it is going out of range and causing the ECU to go into fault? Seems strange that it should be coupled to a change in input.

**Reply by black-top** [reply]
I will try it after work. Just an idea for the future, small I2C LCD oled or text, for basic information display, IP address, flow bench status, and maybe main measurements also. It will be a lot quicker to diagnose problems, and there are no needs for pcb modifications.

**Reply by DeeEmm** [reply]
The project started off with such a display when it ran on Arduino. You can still see a photo of the screen on the header of the website and Facebook group. The screen was junked with the upgrade to the ESP32 as its WiFi and on board server capabilities are far superior. The general recommendation is to dedicate a cheap tablet to the bench and use a browser.

99% of issues can be diagnosed via the serial monitor. Pretty much most signal paths / values can be interrogated there. Typing '?' will return a list of commands. Generally once the bench is commissioned there should be no need to run the serial monitor again, but it is a very useful diagnostics tool and can easily be modded to interrogate any data you want.

The serial monitor link / HTTP schema is also designed to be an API. So it is possible to use ether the serial or HTTP schema to connect third party  or custom devices and control / interrogate the ECU. This way it is possible to build an external device such as an Arduino with OLED display to show values from the bench. 

There is no plan to provide support for external displays within the DIYFB core code.

**Reply by black-top** [reply]
I was unable to recreate the data freeze with the usb cable connected and the serial monitor running. A few times the screen was frozen without the serial monitor connected, but when I connect the usb and opened the serial monitor, the connection was restarted and everything was working fine.  I get no errors in the buffer. Probably some issue with wifi? 

What I noticed that if usb cable is connected I get noise readings in depression, air flow and sometimes in pitot readings. As soon as usb cable disconnected from pc, everything goes to 0 as it should.  I'm considering to put the pcb in metal enclosure for shielding and try to connect external wifi antenna, unfortunately my esp32 board do not have connector for external antenna, but I saw some people successfully modified pcb antenna to solder external antenna directly.   

The main issue with fluctuating readings was solved with piping before MAF and these small issues can be ignored for now.

I think the only problem for further testing right now is problem with leak and flow calibrations.

**Reply by DeeEmm** [reply]
You can solder a wire and a balun connector so that you can connect a regular wifi aerial to the board. It is generally pretty straight forwards. I do the same on other projects.

The USB interference issue is likely exactly that. Interference introduced via the USB cable. This can be one of two differenet forms..

1. Noise from the PSU of the computer entering the ESP via the 5v bus
2. EMI from the environment

1. Is unlikely although not impossible. Generally PC PSUs are pretty good quality as PCs also do not like interference. If you have a laptop, it is also less likely as the battery acs like a giant capacitor and smoothes out transient ripples. Large spikes however will still  get through as they may exceed the systems ability to absorb them.

2. This is most likely. In a purely electronic system EMI is unusual, except for EMI sensitive devices. But as we are also using mains voltages near to the device, the risk increases. Electric motors are particularly noisy from an EMI perspective. All field wiring is potentially vulnerable, not only will it act like an aerial, but all field devices are active and take power from the 5v bus. This means that any EMI on the field wiring will find its way back to the 5v bus. 

One simple solution is to shield all field wiring. But the shield on its own is not enough. One end of the shield needs to be tied to earth, if it is not, the shield will have no effect, as there is no path for the EMI to dissipate. 

Shielding the ECU in a faraday cage (metal box) may help with any EMI acting directly on the ECU. However it is much more likely that any EMI will be introduced via the field wiring directly onto the 5V bus - bypassing the faraday cage altogether.

HTH

**Reply by black-top** [reply]
Maf wire is already shielded, and all other sensors are directly on the board, so only the power supply wire is unshielded right now.  But I see that noise even with the pump and vfd off, and only with usb connected, that's probably from a laptop power supply, it is an old computer dedicated to flow bench, the battery is not working anymore and probably the filtering capacitors in the power supply are dried out. But usb is only for troubleshooting, so no big deal. Just if the board is sensitive, a metal enclosure might help, because it is difficult to tell what the source of small fluctuations is on the running flow bench.

**Reply by DeeEmm** [reply]
Haha yes. I've got a stack of old laptops like that lol.

Noise from the laptop is of course possible

All signal inputs have basic RC filtration on them, even on the board you currently have, but of course it is possible that it could be better. Filtering on the power bus side of things is largely reliant on the caps on the buck PSUs, so there may be room for improvement there too. With the USB connected directly to the ESP perhaps adding a cap at the 5v pin at the PCB mount might help

One change that I've made with the new PCBs is to add a copper ground plane. This will definitely help as well.

**Reply by black-top** [reply]
Does it make sense to upgrade the version V.2.0-RC.6 Build: 23041301 to the current one, or are the changes only for orifices flow bench type?  "This includes a minor fix for the baro display units in the GUI. " this affects only display or calculations too?

**Reply by DeeEmm** [reply]
The master is the latest version. It includes the changes for the differential pressure. The ~latest version~ current release does not include those changes but does include some other fixes, which are listed on the release page - https://github.com/DeeEmm/DIY-Flow-Bench/releases

At this stage I would advise to always use the current ~version~ release. Mostly as at this stage in development, releases are not feature release but, bugfix releases.

NOTE: Edited for clarity

**Reply by black-top** [reply]
Is there any chance to take a look at calibration issues in the near future? I think I would wait for that update.

**Reply by DeeEmm** [reply]
I've created issue #178 for it. I'll try to take a look over the next couple of days. Totally flat out at the moment with other commitments so very time poor ATM.

**Reply by DeeEmm** [reply]
Also created #179 for the flow calibration issue.

Have uploaded fixes to master. Please test and reply in issue thread.

**Comment by stefan63** [comment]
That was the reason I suggested that the code only locks during the
messurements.
When you pause and prepair for next cylinder head section it will be
enabled and take new readings for next messurement.
I asume that the enviromental sensor is there just to adjust the actual
calculations and air pressure, temp and humidity rarly changes super quick.
Messurements would still be direct comparable as the compensation is don
just before the messurements are taken.... I think?


Den ons 20 nov. 2024 kl 13:06 skrev Mick Percy ***@***.***>:

**Comment by stefan63** [comment]
Maybe the button could be named something like "Lock to current BME values"

Den ons 20 nov. 2024 kl 15:40 skrev Stefan Lindgren ***@***.***>:

**Comment by DeeEmm** [comment]
Sorry I am not sold on an additional button to freeze the display. I think it is unnecessary, especially as using it introduces error.

When the current capture button is pressed, it takes a snapshot at an instance in time. i.e. it grabs all of the sensor data within a window of milliseconds, therefore there can be no error in the calculations.

If you freeze some of the sensor values at a point in the past, then they will be out of date when you perform the capture process. So it introduces an error.

Environmental values do change fairly drastically, even over a relatively short timeframe. Much like you can watch a cloud pass overhead, barometric pressure, temperature and humidity can change at the same rate.

The belief that environmental data does not change and can therefore safely be ignored is an artifact from old flow bench sales brochures where they try to justify and downplay the fact that they completely ignore environmental data altogether and had no way to easily capture it. These are the kinds of issues / errors that we are trying to eliminate, the last thing I want to do is reintroduce them.

I think that the main issue in this case is predominantly sensor choice. The 7025 has a range of +/- 25kpa, whereas the ideal range is +/- 7kpa, this means that the resolution of the system is reduced to 25 / 7 = 3.56. So the overall resolution of the system is reduced by 66% as you are effectively only using a third of the sensor range. This means that the smallest change that can be detected is now an order of magnitude higher, which results in seeing larger changes in the GUI. This is the tradeoff for using a sensor with a wider range.

This also assumes that you are running at FSD (full scale deflection of the sensor). i.e. 28". If you run at say 10", then you can further reduce the resolution by a factor of 28/10 = 2.8. 

The effect of this would be most notable in the adjusted flow values as the error is essentially being multipled

I'm not saying that the issue is entirely sensor related, there are of course improvements that can be made to value smoothing for the GUI, which are mapped out in the roadmap, 

Manually freezing the values in the GUI / code is not actually addressing the problem. and will actually make the results less accurate.

**Comment by stefan63** [comment]
Explanation accepted :D
I love how fast this project is running really enjoy following all the
discussions.
Looking forward to put this in a real flow bench.
Meantime I use my ruff test setup and I have now realized how importent it
is to minimize air leaks as this also help to make the live readings much
more stable.

Den tors 21 nov. 2024 kl 01:24 skrev Mick Percy ***@***.***>:
